presto-main/src/main/java/com/facebook/presto/memory/RemoteNodeMemory.java

```diff
@@ -75,19 +87,38 @@ public void asyncRefresh()
             lastWarningLogged.set(System.nanoTime());
         }
         if (sinceUpdate.toMillis() > 1_000 && future.get() == null) {
-            Request request = prepareGet().setUri(memoryInfoUri).build();
+            Request request = preparePost()
+                    .setUri(memoryInfoUri)
+                    .setHeader(CONTENT_TYPE, JSON_UTF_8.toString())
+                    .setBodyGenerator(jsonBodyGenerator(assignmentsRequestJsonCodec, assignments))
+                    .build();
             HttpResponseFuture<JsonResponse<MemoryInfo>> responseFuture = httpClient.executeAsync(request, createFullJsonResponseHandler(memoryInfoCodec));
-            future.set(responseFuture);
+            future.compareAndSet(null, responseFuture);
 
             Futures.addCallback(responseFuture, new FutureCallback<JsonResponse<MemoryInfo>>() {
                 @Override
                 public void onSuccess(@Nullable JsonResponse<MemoryInfo> result)
+
                 {
-                    if (result != null && result.hasValue()) {
-                        memoryInfo.set(ofNullable(result.getValue()));
+                    if (result != null) {
+                        if (result.hasValue()) {
+                            memoryInfo.set(ofNullable(result.getValue()));
+                        }
+                        if (result.getStatusCode() != OK.code()) {
+                            log.warn("Error fetching memory info from %s returned status %d: %s", memoryInfoUri, result.getStatusCode(), result.getStatusMessage());
+                        }
                     }
                     lastUpdateNanos.set(System.nanoTime());
                     future.compareAndSet(responseFuture, null);
+                    long version = currentAssignmentVersion.get();
+                    while (version < assignments.getVersion()) {
```

Comment:
assuming the number only gets larger, you don't need a loop here (but you still need the compare and set)